src/poetry/inspection/lazy_wheel.py

```diff
@@ -0,0 +1,755 @@
+        try:
+            # Initial range request for just the end of the file.
+            file_length, tail = self._try_initial_chunk_request(initial_chunk_size)
+        except HTTPError as e:
+            resp = e.response
+            code = resp.status_code if resp is not None else None
+            # Our initial request using a negative byte range was not supported.
+            if code in [codes.not_implemented, codes.method_not_allowed]:
+                # pypi notably does not support negative byte ranges: see
+                # https://github.com/pypi/warehouse/issues/12823.
+                logger.debug(
+                    "Negative byte range not supported for domain '%s': "
+                    "using HEAD request before lazy wheel from now on",
+                    domain,
+                )
+                # Avoid trying a negative byte range request against this domain for the
+                # rest of the resolve.
+                self._domains_without_negative_range.add(domain)
+                # Apply a HEAD request to get the real size, and nothing else for now.
+                return (self._content_length_from_head(), None)
+            # This indicates that the requested range from the end was larger than the
+            # actual file size: https://www.rfc-editor.org/rfc/rfc9110#status.416.
+            if code == codes.requested_range_not_satisfiable:
+                # In this case, we don't have any file content yet, but we do know the
+                # size the file will be, so we can return that and exit here.
+                assert resp is not None
+                file_length = self._parse_full_length_from_content_range(
+                    resp.headers["Content-Range"]
+                )
+                return (file_length, None)
+            # If we get some other error, then we expect that non-range requests will
+            # also fail, so we error out here and let the user figure it out.
+            raise
+
+        # Some servers that do not support negative offsets,
+        # handle a negative offset like "-10" as "0-10".
+        if int(
+            tail.headers["Content-Length"]
+        ) == initial_chunk_size + 1 and tail.headers["Content-Range"].startswith(
+            f"bytes 0-{initial_chunk_size}"
+        ):
+            tail = None
+            self._domains_without_negative_range.add(domain)
+        return file_length, tail
+
+    def _prefetch_metadata(self, name: str) -> str:
+        """Locate the *.dist-info/METADATA entry from a temporary ``ZipFile`` wrapper,
+        and download it.
+
+        This method assumes that the *.dist-info directory (containing e.g. METADATA) is
+        contained in a single contiguous section of the zip file in order to ensure it
+        can be downloaded in a single ranged GET request."""
+        logger.debug("begin prefetching METADATA for %s", name)
+
+        metadata_regex = re.compile(r"^[^/]*\.dist-info/METADATA$")
```

Comment:
Is there any reason not to compile this earlier and have this pattern as a constant?

Reply:
No. Changed it.